vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java

```diff
@@ -3004,11 +3013,39 @@ private void setSortOrder(List<GridSortOrder<T>> order,
         Objects.requireNonNull(order, "Sort order list cannot be null");
 
         if (sortOrder.equals(order)) {
+            if (!order.isEmpty()) {
+                updateClientSideSorterIndicators(order);
+            }
             return;
         }
 
+        boolean isOldValSameNewValWithoutOrder = false;
+        List<GridSortOrder<T>> tempSortOrder = sortOrder.stream().collect(Collectors.toList());
+        if (tempSortOrder.containsAll(order) && order.containsAll(tempSortOrder)) {
+            isOldValSameNewValWithoutOrder = true;
+        }
+
         if (!userOriginated) {
             updateClientSideSorterIndicators(order);
+        } else if (userOriginated
+                && !sortOrder.isEmpty()
+                && !order.isEmpty()
+                && sortOrder.size() == order.size()
+                && isOldValSameNewValWithoutOrder) {
+            order = sortOrder.stream().collect(Collectors.toList());
+            updateClientSideSorterIndicators(order);
+            return;
+        }
+
+        boolean isSortOrderNotEmptyAfterReAttached = false;
+        if (order.isEmpty() && isSorted && !sortOrder.isEmpty() && !isPreviousSortValueNotEmpty) {
+            isSortOrderNotEmptyAfterReAttached = true;
+        }
+        // In case that reattach the grid or something to keep the sort-state
+        // exception cases: reset, or set back to sort size = 0
+        if (!isResetSort && isSortOrderNotEmptyAfterReAttached) {
+            order = sortOrder.stream().collect(Collectors.toList());
+            updateClientSideSorterIndicators(order);
```

Comment:
`updateClientSideSorterIndicators` is now invoked in 4 different places with different conditions within the same method! Does that 1 added test case really cover all of these cases? You should rethink your approach to fixing this issue as this is getting really complex.